envoy/certificate_provider/certificate_provider.h

```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+#include <list>
+
+#include "envoy/common/callback.h"
+#include "envoy/common/pure.h"
+
+#include "absl/strings/string_view.h"
+#include "openssl/ssl.h"
+#include "openssl/x509v3.h"
+
+namespace Envoy {
+namespace CertificateProvider {
+
+struct Certpair {
+  const std::string& certificate_;
+  const std::string& private_key_;
+};
+
+/**
+ * This is used to implement asynchronous certificate provider.
+ */
+class CertificateSubscriptionCallbacks {
```

Comment:
How will this class to be implemented and called? I don't see a usage. Also comment for every method.

Reply:
See the new staticCertificateProvider as an example, it supports generating certs so "static" is not good, I can rename it later.

Reply:
@lizan  @markdroth  I rename it back to DefaultCertificateProvider, it shows how the asynchronous interfaces are used. Please have a look.

Reply:
Ha, you're replying on the old patch, could you review the latest code? Currently we have an addUpdateCallback, it serves like startWatch I think, maybe we need a delUpdateCallback but I'm not sure if we really need it.

Reply:
I did look at the latest patch.  My comments are based on that.

Having `addUpdateCallback()` helps, but you still need to separate out `CertificateSubscriptionCallbacks` from `CertificateProvider`.  I think an instance of `CertificateSubscriptionCallbacks` should be passed to `addUpdateCallback()`.

Reply:
Please see my comment on the latest code.